_posts/2024-05-11-coroutines-and-deadlocks.md

```diff
@@ -0,0 +1,231 @@
+}
+```
+
+In the example above, `delay(100)` is a suspending function. 
+When this line is executed, the coroutine handling this block of code is suspended, releasing the thread to perform other tasks. 
+This behavior becomes problematic due to how JDBC and database connections operate.
+
+## Scenario Leading to Deadlock
+
+JDBC operations are inherently blocking. 
+When a transaction is initiated with JDBC, it holds onto a database connection from a connection pool until the transaction is complete.
+Connection pools, typically configured with a maximum number of connections (default around 10), are used to manage these expensive resources efficiently.
+
+Consider a scenario where you're handling a high volume of simultaneous requests that interact with the database:
+
+1. Let's say there are 100 simultaneous requests and a Dispatchers.IO context configured with 64 threads. 
+These threads begin processing the requests, each requiring a database connection to perform the transaction.
+2. As the first 10 coroutines, processing the requests, grab connections and proceed, they hit a suspend function and suspend. The threads are freed up and pick up the next set of requests.
+3. Eventually, all available database connections are allocated, but they're held by suspended coroutines that cannot proceed because there are no threads available to continue their execution. 
+At the same time, no additional connections are available for new coroutines that are scheduled, leading to a deadlock. The threads are idle, waiting for connections to free up, and the connections are locked in suspended transactions that can't proceed without a thread.
+
+This situation outlines a classical deadlock scenario where both resources (threads and database connections) are fully utilized but in a state that prevents any further progress.
+
+
+
+## Conclusion of Understanding the Risks of Mixing Exposed with Coroutines
+
+So, how do we overcome the deadlock issue? 
+
+One possible workaround would be to increase the number of database connections to exceed the maximum number of Dispatcher.IO threads. 
+However, this approach is costly and inefficient, as the number of database connections provided by the databases is limited. 
+This becomes especially problematic in environments with multiple replicas of your application, 
+which would exponentially increase the required connections. 
+Furthermore, our physical database is shared by multiple services, making the necessary number of connections unfeasible.
+
+Another idea was to avoid calling suspending functions within a coroutine when using Exposed. 
+However, this brings up an important question: Why use coroutines with Exposed at all if you can't utilize the feature to call other suspending functions within a transaction? 
+This limitation highlights a fundamental challenge with the current design: it doesn't allow a feasible solution for fully leveraging coroutines in database operations.
+
```

Comment:
I do not get it how? At some point in time you have to access the database which, with JDBC, is a blocking operation.

Reply:
a message passing architecture would be completely different, essentially trying to reduce shared mutable state, maybe not even using coroutine multithreading. but maybe this can only solve race conditions, but not deadlocks.

i just wondered if there can be a strict mechanism that prevents deadlocks 100%, just like rust prevents data races 100%.